robottelo/containers.py

```diff
@@ -0,0 +1,249 @@
+import logging
+
+from uuid import uuid4
+from robottelo.config import settings
+
+logger = logging.getLogger(__name__)
+
+
+class ContainerError(Exception):
+    """Exception raised for failed container management operations."""
+
+
+class DockerContainer:
+    """
+    Gather information needed to spin up a Docker-based container.
+    Attributes:
+            image (init): The name of the container image to build from.
+            tag (init): The tag of the operating system to use.
+            agent (boolian): If set to True, prepares container for
+            katello-agent.
+    """
+
+    def __init__(self, image, tag, agent, ports):
+        """
+        Parameters:
+            image (init): The name of the container image to build from.
+            tag (init): The tag of the operating system to use.
+            agent (boolian): If set to True, prepares container for
+            katello-agent.
+        Returns:
+            A container object if tag and image supplied.
+        """
+        import docker
+        import docker.api.container
+
+        self._client = docker.Client(version="1.22")
+        self.image = image
+        self.tag = tag
+        self.name = uuid4()
+        self.ports = ports or {}
+        self._inst = None
+        self._logs = {}
+        self._mount = agent
+        self._create()
+
+    def _create(self):
+        """Creates containers for an instance of this class.
+        Mounts /dev/log as a read-write file system if katello-agent
+        is required.
+        Logs RHSM logs to stout.
+        Logs host name to stout after starting the container.
+        """
+        logging.info(f"Creating {self.tag} container named {self.name}")
+        volumes = (
+            {"/dev/log": {"bind": "/dev/log", "mode": "rw"}} if self._mount
+            else {}
+        )
+        self._inst = self._client.create_container(
+            detach=False,
+            host_config=self._client.create_host_config(
+                binds=volumes, port_bindings=self.ports
+            ),
+            image=f"{self.image}:{self.tag}",
+            ports=list(self.ports.keys()),
```

Comment:
I'm not able to understand how is the port will be used here. can you please explain here with an example

Reply:
docker-py is a bit wonky. not only does it require a port_bindings mapping (above), but also explicitly listing the ports (internal to the container) that need to be exposed in this argument.
so the port mapping is internal:external while this argument is just the internal ports.